### ISSUE TITLE
reporting callback

### ISSUE THREAD

**Issue by azu** [fiznool/express-mongo-sanitize#35]
I want to get a callback option that is called after sanitizing the request.

For example, `report` callback is called if the request is sanitized.

```js
app.use(mongoSanitize({
  /**
   * @param {*} req request object
   * @param {string} key sanitized request property name(header,body,params,query)
   * @param {*} value sanitized value
   **/
  report({ req, key, value }){
    console.warn("This request is sanitized. please check the usage", req, key, value);
  }
));
```

Another option: allow to pass function to `replaceWith` option.

```js
// dry-run and report
app.use(mongoSanitize({
  /**
   * @param {*} value
   **/
  replaceWith: (value) => {
    console.warn("This request is sanitized. please check the usage", value);
    return value;
  }
));
// drop and report
app.use(mongoSanitize({
  /**
   * @param {*} value
   **/
  replaceWith: (value) => {
    console.warn("This request is sanitized. please check the usage", value);
    return; // if return undefined, drop the property
  }
));
// replace and report
app.use(mongoSanitize({
  /**
   * @param {*} value
   **/
  replaceWith: (value) => {
    console.warn("This request is sanitized. please check the usage", value);
    return "_";
  }
));
```

## Use Case

Currently, [express-mongo-sanitize](https://github.com/fiznool/express-mongo-sanitize) remove or replace the invalid properties silently.
I think that the callback API helps us in the following cases.

- Logging attacker payload
- Show warning to the wrong usage of developers

Do you think about this?

---

FYI: I've created [express-mongo-sanitize-reporter](https://github.com/azu/express-mongo-sanitize-reporter) that is wrapper of [express-mongo-sanitize](https://github.com/fiznool/express-mongo-sanitize).
It aims to run as **dry-run** and report invalid requests for warning before introducing express-mongo-sanitize.
It is good that express-mongo-sanitize has callback API, I think.

**Comment by fiznool** [comment]
Thank you for this. I think this is a good idea!

I quite like the idea of keeping the `replaceWith` property simple, as you proposed with the `report` callback - but I think that `onSanitize` is a more descriptive property name. It would also be nice to have a dry run mode. 

With this in mind, allow me to propose something slightly different. Two new options:

- `onSanitize`: this is the same as the `report` callback above, accepting a single object with `req, key, value` properties. Although, perhaps we should replace `value` with `originalValue`, and set a `sanitizedValue` prop if `replaceWith` option was used?
- `dryRun`: optional boolean. Setting to `true` will not sanitize the key, instead it will log out to the console, something along the lines of:
  - (Remove mode) `Dry Run: key ${key} with value ${originalValue} would be removed 
  - (Preserve mode) `Dry Run: key ${key} would be changed from ${originalValue} to ${sanitizedValue}

When reporting on `key` we will need to take into account nested keys, and should probably report these in dotted notation, e.g. `nested.key`.

What do you think?

**Comment by azu** [comment]
> I think that onSanitize is a more descriptive property name. 

Yes. I agree.

> onSanitize: this is the same as the report callback above, accepting a single object with req, key, value properties. Although, perhaps we should replace value with originalValue, and set a sanitizedValue prop if replaceWith option was used?

Sure.

Minimal requirement is following.

- `req`  is exrpress's request object
- `key`(more reasonable name?) is one of `'body', 'params', 'headers', 'query'`

The use can access the value via `req[key]` that is sanitized value.

> * `dryRun`: optional boolean. Setting to `true` will not sanitize the key, instead it will log out to the console, something along the lines of:

It looks like Good.

It maybe good that add prefix like `[express-mongo-sanitize]` for searchablility.

> [express-mongo-sanitize] Dry Run: key ${key} with value ${originalValue} would be removed

**Comment by fiznool** [comment]
I think we have a plan. Would you be willing to submit a PR which makes these changes?

**Comment by azu** [comment]
OK. I will try it.

I noticed that we do not need to show the console log if dryRun is enabled.
The user can use `onSanitize` instead of the built-in log.

```js
app.use(mongoSanitize({
  dryRun: true,
  onSanize({ req, key }){
    console.warn("This request is not sanized because it is dryRun mode. However, please check the usage", req, key);
  }
));
```

**Comment by azu** [comment]
I've implemented in #39
